### PR TITLE
[SPARK-38385][SQL] Improve error messages of 'mismatched input' cases from ANTLR

### PR DESCRIPTION
**Pull request by anchovYu**
### What changes were proposed in this pull request?
This PR handles the case 1 mentioned in https://issues.apache.org/jira/browse/SPARK-38385:
* Before
    ```
    ParseException: 
    mismatched input 'sel' expecting {'(', 'APPLY', 'CONVERT', 'COPY', 'OPTIMIZE', 'RESTORE', 'ADD', 'ALTER', 'ANALYZE', 'CACHE', 'CLEAR', 'COMMENT', 'COMMIT', 'CREATE', 'DELETE', 'DESC', 'DESCRIBE', 'DFS', 'DROP', 'EXPLAIN', 'EXPORT', 'FROM', 'GRANT', 'IMPORT', 'INSERT', 'LIST', 'LOAD', 'LOCK', 'MAP', 'MERGE', 'MSCK', 'REDUCE', 'REFRESH', 'REPLACE', 'RESET', 'REVOKE', 'ROLLBACK', 'SELECT', 'SET', 'SHOW', 'START', 'SYNC', 'TABLE', 'TRUNCATE', 'UNCACHE', 'UNLOCK', 'UPDATE', 'USE', 'VALUES', 'WITH'}(line 1, pos 0)
    
    == SQL ==
    sel 1
    ^^^ 
    ```
* After
    ```
    ParseException: 
    syntax error at or near 'sel'(line 1, pos 0)
    
    == SQL ==
    sel 1
    ^^^ 
    ```

#### Implementation general idea
ANTLR uses the DefaultErrorStrategy class to create error messages: 

```scala
public class DefaultErrorStrategy implements ANTLRErrorStrategy {
  protected void reportInputMismatch(Parser recognizer, InputMismatchException e)
  {
	String msg = "mismatched input " + 
                    getTokenErrorDisplay(e.getOffendingToken()) + " expecting " +
                    e.getExpectedTokens().toString(recognizer.getVocabulary());
	recognizer.notifyErrorListeners(e.getOffendingToken(), msg, e);
  }
  ..
}
```
It is easy to extend the `DefaultErrorStrategy` and override corresponding functions to output better error messages. Then in our parser, set the error strategy to be the one we created.

#### Changes in code
To achieve this, the following changes are made:
* error-classes.json
    Define a new type of error `PARSE_INPUT_MISMATCHED` with the new error framework:
    ```json
      "PARSE_INPUT_MISMATCHED" : {
        "message" : [ "syntax error at or near %s" ],
        "sqlState" : "42000"
      },
    ```
* SparkParserErrorStrategy.scala
    This is a new class, extending the `org.antlr.v4.runtime.DefaultErrorStrategy` that does special handling on errors. Note the original `DefaultErrorStrategy` is where the `mismatched input` error message generates from. 
    The new class is intended to provide more information, e.g. the error class and the message parameters, on these errors encountered in ANTLR parser to the downstream consumers to be able to apply the `SparkThrowable` error message framework to these exceptions.

* ParserDriver.scala
  * It sets the error strategy of the parser to be the above new `SparkParserErrorStrategy`.
  * When catching an exception thrown from ANTLR, when it can find out the error class and message parameter, it creates `ParseException` with this information, which composes the error message through `SparkThrowableHelper.getMessage`. It then formalizes the standard error messages of these types.

* test suites
    It change all affected test suites. It also adds a check on the error class, note the newly added `PARSE_INPUT_MISMATCHED` in the after case:
    ```scala
    // before
    intercept("select * from r order by q from t", 1, 27, 31,
      "mismatched input",
      "---------------------------^^^")
    // after
    intercept("select * from r order by q from t", "PARSE_INPUT_MISMATCHED",
      1, 27, 31,
      "syntax error at or near",
      "---------------------------^^^"
    )
    ```

### Why are the changes needed?
https://issues.apache.org/jira/browse/SPARK-38384 The description states the reason for the change.
TLDR, the error messages of ParseException directly coming from ANTLR are not user-friendly and we want to improve it.

### Does this PR introduce _any_ user-facing change?
If the error messages change are considered as user-facing change, then yes.
The changes in the error message:
1. Adjust the words, from ‘mismatched input {}’ to a more readable one, ‘syntax error at or near {}.’. This also aligns with the PostgreSQL error messages. 
2. Remove the expecting full list.

One example case is listed in the top of this PR description.

### How was this patch tested?
Through manual local tests.

### CODE REVIEW

**Comment by anchovYu** [comment]
@MaxGekk @cloud-fan would you take a look? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/SparkParserErrorStrategy.scala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * A [[SparkRecognitionException]] extends the [[RecognitionException]] with more information
+ * including the error class and parameters for the error message, which align with the interface
+ * of [[SparkThrowableHelper]].
+ */
+private[parser] class SparkRecognitionException(
+    message: String,
+    recognizer: Recognizer[_, _],
+    input: IntStream,
+    ctx: ParserRuleContext,
+    val errorClass: Option[String] = None,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/SparkParserErrorStrategy.scala:35]: are we going to pass error classes other than `PARSE_INPUT_MISMATCHED` to it?

Comment (anchovYu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/SparkParserErrorStrategy.scala:35]: Yes, the `PARSE_INPUT_MISMATCHED` is the first step. I have the plan to improve some other error messages with new error class attached.

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -122,6 +122,10 @@
     "message" : [ "PARTITION clause cannot contain a non-partition column name: %s" ],
     "sqlState" : "42000"
   },
+  "PARSE_INPUT_MISMATCHED" : {
+    "message" : [ "syntax error at or near %s" ],
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:126]:
```suggestion
    "message" : [ "Syntax error at or near %s" ],
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/SparkParserErrorStrategy.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+import org.antlr.v4.runtime.{DefaultErrorStrategy, InputMismatchException, IntStream, Parser,
+  ParserRuleContext, RecognitionException, Recognizer}
+
+/**
+ * A [[SparkRecognitionException]] extends the [[RecognitionException]] with more information
+ * including the error class and parameters for the error message, which align with the interface
+ * of [[SparkThrowableHelper]].
+ */
+private[parser] class SparkRecognitionException(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/SparkParserErrorStrategy.scala:30]:
```suggestion
class SparkRecognitionException(
```
everything in catalyst is private, we don't need to add private modifiers.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/SparkParserErrorStrategy.scala [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  /** Construct from a given [[RecognitionException]], with additional error information. */
+  def this(
+      recognitionException: RecognitionException,
+      errorClass: String,
+      messageParameters: Array[String]) =
+    this(
+      recognitionException.getMessage,
+      recognitionException.getRecognizer,
+      recognitionException.getInputStream,
+      Try { recognitionException.getCtx.asInstanceOf[ParserRuleContext] } match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/SparkParserErrorStrategy.scala:48]: how about
```
recognitionException.getCtx match {
  case p: ParserRuleContext => p
  case _ => null
}
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala [hunk 5]
```diff
@@ -1774,7 +1777,7 @@ class DDLParserSuite extends AnalysisTest {
         allColumns = true))
 
     intercept("ANALYZE TABLE a.b.c COMPUTE STATISTICS FOR ALL COLUMNS key, value",
-      "mismatched input 'key' expecting {<EOF>, ';'}")
+      Some("PARSE_INPUT_MISMATCHED"), "syntax error at or near 'key'") // expecting {<EOF>, ';'}
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1780]: `expecting {<EOF>, ';'}` is not in the error message any more.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ErrorParserSuite.scala [hunk 6]
```diff
@@ -31,26 +32,47 @@ class ErrorParserSuite extends AnalysisTest {
...
   }
 
-  def intercept(sqlCommand: String, messages: String*): Unit =
-    interceptParseException(CatalystSqlParser.parsePlan)(sqlCommand, messages: _*)
-
-  def intercept(sql: String, line: Int, startPosition: Int, stopPosition: Int,
-                messages: String*): Unit = {
+  private def interceptImpl(sql: String, messages: String*)(
+    line: Option[Int] = None, startPosition: Option[Int] = None, stopPosition: Option[Int] = None)(
+    errorClass: Option[String] = None): Unit = {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ErrorParserSuite.scala:37]:
```suggestion
    line: Option[Int] = None,
    startPosition: Option[Int] = None, 
    stopPosition: Option[Int] = None,
    errorClass: Option[String] = None): Unit = {
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ErrorParserSuite.scala [hunk 7]
```diff
@@ -31,26 +32,47 @@ class ErrorParserSuite extends AnalysisTest {
...
+    }
+
+    // Check error class.
+    if (errorClass.isDefined) {
+      assert(e.getErrorClass == errorClass.get)
+    }
+  }
+
+  def intercept(sqlCommand: String, errorClass: Option[String], messages: String*): Unit = {
+      interceptImpl(sqlCommand, messages: _*)()(errorClass)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ErrorParserSuite.scala:64]: nit: 2 spaces indentation

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ErrorParserSuite.scala [hunk 8]
```diff
@@ -31,26 +32,47 @@ class ErrorParserSuite extends AnalysisTest {
...
+  }
+
+  def intercept(sql: String, line: Int, startPosition: Int, stopPosition: Int,
+    messages: String*): Unit = {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ErrorParserSuite.scala:68]:
```suggestion
  def intercept(
      sql: String, line: Int, startPosition: Int, stopPosition: Int, messages: String*): Unit = {
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala (hunk 5 above)
Comment (anchovYu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:1780]: I make it a comment in case in the future we are able to provide suggestion, and that may serves a reminder of what are really expected in the pre-changed code.
But I can remove it anyways.

**Comment by anchovYu** [comment]
Thanks @cloud-fan for the review. I feel sorry for all the linting problem.. Before I commit, I ran the dev/scalastyle and it shows all style has passed, then I thought it is fine. So to avoid this happen again next time and to confirm, this scalastyle check is not enough right? We should still follow the code guidance manually?

**Comment by cloud-fan** [comment]
Yea we don't have a very good style detection script. In practise we follow https://github.com/databricks/scala-style-guide

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala [hunk 9]
```diff
@@ -200,11 +200,15 @@ trait AnalysisTest extends PlanTest {
     }
   }
 
-  protected def interceptParseException(
-      parser: String => Any)(sqlCommand: String, messages: String*): Unit = {
+  protected def interceptParseException(parser: String => Any)(
+    sqlCommand: String, messages: String*)(
+    errorClass: Option[String] = None): Unit = {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala:205]: nit: in other places, we just add the `errorClass` parameter before the `messages` parameter, e.g. `def intercept(sqlCommand: String, errorClass: Option[String], messages: String*)`, but why do we add a curry constructor here?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala (hunk 9 above)
Comment (anchovYu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala:205]: The intention is to provide default value `None` for the errorClass and potential future added fields like sqlState. Especcially if new fields are added, this interface can simply be extended to 
```scala
  protected def interceptParseException(parser: String => Any)(
    sqlCommand: String, messages: String*)(
    errorClass: Option[String] = None, newField: Option[String] = None):
```
and the caller doesn't need to change its signature. For example, 
``` scala
  private def intercept(sqlCommand: String, messages: String*): Unit =
    interceptParseException(parsePlan)(sqlCommand, messages: _*)()

  private def intercept(sqlCommand: String, errorClass: Option[String], messages: String*): Unit =
    interceptParseException(parsePlan)(sqlCommand, messages: _*)(errorClass)
```
will still compile.

Comment (anchovYu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala:205]: WDYT? Is there another set of designs in the consideration of easy extension?

**Comment by cloud-fan** [comment]
thanks, merging to master!
